staging/src/k8s.io/apiserver/pkg/reconcilers/peer_endpoint_lease.go

```diff
@@ -0,0 +1,313 @@
+	"k8s.io/apiserver/pkg/features"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/apiserver/pkg/storage"
+	"k8s.io/apiserver/pkg/storage/storagebackend"
+	storagefactory "k8s.io/apiserver/pkg/storage/storagebackend/factory"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+)
+
+const (
+	APIServerIdentityLabel = "apiserverIdentity"
+)
+
+// PeerLeases is an interface which assists in managing the set of active servers's endpoint leases
+type PeerLeases interface {
+	// ListLeases retrieves a list of the current server IPs
+	ListLeases() ([]string, error)
+
+	// UpdateLease adds or refreshes a servers lease
+	UpdateLease(ip string, peerBindAddress string, serverId string, endpointPorts []corev1.EndpointPort) error
+
+	// GetLease retrieves the IP of a specific service
+	GetLease(serverId string) (string, error)
+
+	// RemoveLease removes a server's lease
+	RemoveLease(ip string, endpointPorts []corev1.EndpointPort) error
+
+	// Destroy cleans up everything on shutdown.
+	Destroy()
+}
+
+type PeerEndpointLeases struct {
+	Storage   storage.Interface
+	DestroyFn func()
+	BaseKey   string
+	LeaseTime time.Duration
+}
+
+// ListLeases retrieves a list of the current server IPs from storage
+func (s *PeerEndpointLeases) ListLeases() ([]string, error) {
+	storageOpts := storage.ListOptions{
+		ResourceVersion:      "0",
+		ResourceVersionMatch: metav1.ResourceVersionMatchNotOlderThan,
+		Predicate:            storage.Everything,
+		Recursive:            true,
+	}
+	ipInfoList, err := s.GetIpInfoList(storageOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	ipList := make([]string, 0, len(ipInfoList.Items))
+	for _, ip := range ipInfoList.Items {
+		if len(ip.Subsets) > 0 && len(ip.Subsets[0].Addresses) > 0 && len(ip.Subsets[0].Addresses[0].IP) > 0 {
+			ipList = append(ipList, ip.Subsets[0].Addresses[0].IP)
+		}
+	}
+
+	klog.V(6).Infof("Current server IPs listed in storage are %v", ipList)
+
+	return ipList, nil
+}
+
+// UpdateLease resets the TTL on a server IP in storage
+// UpdateLease will create a new key if it doesn't exist.
+// We use the first element in endpointPorts as a part of the lease's base key
+// This is done to support out tests that simulate 2 apiservers running on the same ip but
+// different ports
+
+// It will also do the following if UnknownVersionInteroperabilityProxy feature is enabled
+// 1. store the apiserverId as a label
+// 2. store the values passed to --peer-bind-ip and --peer-bind-port flags to kube-apiserver as an annotation
+// with value of format <ip:port>
+func (s *PeerEndpointLeases) UpdateLease(ip string, peerBindAddress string, serverId string, endpointPorts []corev1.EndpointPort) error {
+	key := path.Join(s.BaseKey, ip)
+	if len(endpointPorts) > 0 {
+		key = path.Join(key, fmt.Sprint(endpointPorts[0].Port))
+	}
+	return s.Storage.GuaranteedUpdate(apirequest.NewDefaultContext(), key, &corev1.Endpoints{}, true, nil, func(input kruntime.Object, respMeta storage.ResponseMeta) (kruntime.Object, *uint64, error) {
```

Comment:
Yup, I hate it this way.  Direct to etcd storage requests that I honestly don't remember the gotchas of.

Reply:
The other alternative was to store this directly in apiserver identity lease (coordinationv1.Lease) but wanted to make sure we address @liggitt's concerns around making apiserver rely on an api resource to fetch this info) 

cc @aojea

Reply:
Yes, that is why I suggested to use only the existing v1.Lease object created by the `APIServiceIdentity`controller https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/controlplane/instance.go#L559-L577 , we already need the identity and the information is duplicated on the hidden object that as David says is not great ...

Reply:
So [we decided](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/4020-unknown-version-interoperability-proxy/README.md#identifying-destination-apiservers-network-location) to perform dual writes of the ip info in 

1. storage lease (which will be referenced by the apiservers to find out peer ip info)
2. v1.Lease (which can be used by user to find out what ip is being used by UVIP feature)

we decided to make apiservers use the storage lease since these objects are opaque to the user and there was no risk of the user modifying these objects for malicious reasons. The other nice things to use storage leases were (as @liggitt  had pointed out earlier) `IPs are already there, these are the same IPs populated into the kubernetes service endpoints, API servers already write to it, and is not exposed in a REST API users can spoof`

more detailed pros vs cons list of both approaches are also listed in the [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/4020-unknown-version-interoperability-proxy/README.md#alternatives)

But then eventually we decided to have a new reconciliation loop for endpoint leases since we did not want to modify the legacy endpoint reconciliation code too much. 

I do understand that since we created a new reconciliation loop for endpoint leases, the point about `Existing code to handle lifecycle of the masterleases is convenient` & `IPs being there already` no longer stands, but I want to make sure we address the other security concern around server side forgeries. If that is something that is not a hard requirement, I can switch back to using v1.Lease (for sure, that approach is simpler to implement and understand)

Reply:
To mitigate risk:

- We will not read/write these objects when the feature gate is off (I've added a comment about this).
- Since this etcd lease handling code is mostly cribbed from [reconcilers.go](https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/controlplane/reconcilers/reconcilers.go), [lease.go](https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/controlplane/reconcilers/lease.go) and [controller.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/controlplane/controller/kubernetesservice/controller.go), I've done a review pass comparing the code here with the code it was cribbed from and made sure it looks reasonable.
- We will use serverID as the key here, this makes the approach more consistent with v1.Leases (which also uses serverID), and differs from the other etcd lease which uses IP. This also means these objects are potentially useful for other server identity purposes that the old etcd leases would be insufficient for.